File: LLMgeneratedCode/4o-2.R
library(tidyverse)
library(glmnet)
library(pROC)

# Load data
training_features <- read.csv("training_species_abundance.csv")
validation_features <- read.csv("validation_species_abundance.csv")
training_metadata <- read.csv("training_metadata.csv")
validation_metadata <- read.csv("validation_metadata.csv")

# Ensure common features
common_features <- intersect(names(training_features), names(validation_features))
common_features <- setdiff(common_features, "specimen")

# Filter features
training_features <- training_features[, c("specimen", common_features)]
validation_features <- validation_features[, c("specimen", common_features)]

# Merge metadata
training_data <- merge(training_metadata, training_features, by = "specimen")
validation_data <- merge(validation_metadata, validation_features, by = "specimen")

# Function to select last specimen before a given collect_wk for each participant
filter_last_specimen <- function(data, wk_threshold) {
  data %>%
    filter(collect_wk < wk_threshold) %>%
    group_by(participant_id) %>%
    filter(collect_wk == max(collect_wk)) %>%
    ungroup()
}

# Prepare datasets for PTB (collect_wk < 32)
training_ptb <- filter_last_specimen(training_data, 32)
validation_ptb <- filter_last_specimen(validation_data, 32)

# Define PTB outcome (delivery_wk < 37)
training_ptb$PTB <- as.factor(training_ptb$delivery_wk < 37)
validation_ptb$PTB <- as.factor(validation_ptb$delivery_wk < 37)

# Prepare model matrices
X_train <- as.matrix(training_ptb[, common_features])
y_train <- as.numeric(training_ptb$PTB)  # Convert to numeric for glmnet
X_test <- as.matrix(validation_ptb[, common_features])
y_test <- as.numeric(validation_ptb$PTB)

set.seed(123)
# Fit LASSO logistic regression
cv_fit <- cv.glmnet(X_train, y_train, family = "binomial", alpha = 1)

# Predict on test set
pred_probs <- predict(cv_fit, newx = X_test, s = "lambda.min", type = "response")

# Compute AUC
roc_obj <- roc(y_test, pred_probs)
auc_value <- auc(roc_obj)
print(paste("AUC for PTB:", auc_value))
write.csv(data.frame(y_test=y_test,y_pred=as.numeric(pred_probs)),file="./out2/4o-2A.R.csv")

# Plot ROC curve
plot(roc_obj, col = "blue", main = "ROC Curve for PTB")

# Repeat for EarlyPTB (collect_wk < 28, delivery_wk < 32)
training_earlyptb <- filter_last_specimen(training_data, 28)
validation_earlyptb <- filter_last_specimen(validation_data, 28)

training_earlyptb$EarlyPTB <- as.factor(training_earlyptb$delivery_wk < 32)
validation_earlyptb$EarlyPTB <- as.factor(validation_earlyptb$delivery_wk < 32)

X_train_early <- as.matrix(training_earlyptb[, common_features])
y_train_early <- as.numeric(training_earlyptb$EarlyPTB)
X_test_early <- as.matrix(validation_earlyptb[, common_features])
y_test_early <- as.numeric(validation_earlyptb$EarlyPTB)

cv_fit_early <- cv.glmnet(X_train_early, y_train_early, family = "binomial", alpha = 1)

pred_probs_early <- predict(cv_fit_early, newx = X_test_early, s = "lambda.min", type = "response")

roc_obj_early <- roc(y_test_early, pred_probs_early)
auc_value_early <- auc(roc_obj_early)
print(paste("AUC for EarlyPTB:", auc_value_early))
write.csv(data.frame(y_test=y_test_early,y_pred=as.numeric(pred_probs_early)),file="./out2/4o-2B.R.csv")

plot(roc_obj_early, col = "red", main = "ROC Curve for EarlyPTB")

File: LLMgeneratedCode/4o-2.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score, roc_curve

# Load microbiome feature data
train_features = pd.read_csv("training_species_abundance.csv")
valid_features = pd.read_csv("validation_species_abundance.csv")

# Load metadata
train_metadata = pd.read_csv("training_metadata.csv")
valid_metadata = pd.read_csv("validation_metadata.csv")

# Ensure only common features are used
common_features = set(train_features.columns).intersection(set(valid_features.columns))
common_features.discard("specimen")  # Remove the linking column
train_features = train_features[["specimen"] + list(common_features)]
valid_features = valid_features[["specimen"] + list(common_features)]

# Merge metadata with feature data
train_data = train_metadata.merge(train_features, on="specimen")
valid_data = valid_metadata.merge(valid_features, on="specimen")

def preprocess_data(metadata, features, collect_wk_threshold, delivery_wk_threshold):
    """Filters and prepares the dataset for modeling."""
    
    # Define PTB outcome
    metadata["PTB"] = (metadata["delivery_wk"] < delivery_wk_threshold).astype(int)
    
    # Select the latest specimen before the threshold for each participant
    metadata = metadata[metadata["collect_wk"] < collect_wk_threshold]
    metadata = metadata.sort_values(by=["participant_id", "collect_wk"], ascending=[True, False])
    metadata = metadata.drop_duplicates(subset=["participant_id"], keep="first")
    
    # Merge with features
    final_data = metadata.merge(features, on="specimen")
    
    # Prepare feature matrix and target variable
    X = final_data.drop(columns=["specimen", "participant_id", "delivery_wk", "PTB"])
    y = final_data["PTB"]
    
    return X, y

# Process data for PTB prediction
X_train, y_train = preprocess_data(train_metadata, train_features, collect_wk_threshold=32, delivery_wk_threshold=37)
X_valid, y_valid = preprocess_data(valid_metadata, valid_features, collect_wk_threshold=32, delivery_wk_threshold=37)

# Train a RandomForest model
clf = RandomForestClassifier(n_estimators=100, random_state=42)
clf.fit(X_train, y_train)

# Evaluate model
y_pred_prob = clf.predict_proba(X_valid)[:, 1]
auc_score = roc_auc_score(y_valid, y_pred_prob)
print(f"AUC-ROC for PTB Prediction: {auc_score:.4f}")

# Plot ROC Curve
fpr, tpr, _ = roc_curve(y_valid, y_pred_prob)
plt.figure()
plt.plot(fpr, tpr, label=f'PTB (AUC = {auc_score:.2f}')
plt.plot([0, 1], [0, 1], 'k--')
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('ROC Curve for PTB Prediction')
plt.legend()
plt.savefig("./out2/4o-2.py_A.pdf", format="pdf")

# Repeat for EarlyPTB (delivery_wk < 32, collect_wk < 28)
X_train, y_train = preprocess_data(train_metadata, train_features, collect_wk_threshold=28, delivery_wk_threshold=32)
X_valid, y_valid = preprocess_data(valid_metadata, valid_features, collect_wk_threshold=28, delivery_wk_threshold=32)

clf.fit(X_train, y_train)
y_pred_prob = clf.predict_proba(X_valid)[:, 1]
auc_score = roc_auc_score(y_valid, y_pred_prob)
print(f"AUC-ROC for EarlyPTB Prediction: {auc_score:.4f}")

# Plot ROC Curve for EarlyPTB
fpr, tpr, _ = roc_curve(y_valid, y_pred_prob)
plt.figure()
plt.plot(fpr, tpr, label=f'EarlyPTB (AUC = {auc_score:.2f}')
plt.plot([0, 1], [0, 1], 'k--')
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('ROC Curve for EarlyPTB Prediction')
plt.legend()
plt.savefig("./out2/4o-2.py_B.pdf", format="pdf")
